File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int[] charIndex = new int[128];
        int maxlength = 0;
        int left = 0;
        Arrays.fill(charIndex , -1);

        for(int right = 0; right < n; right++){
            if(charIndex[s.charAt(right)] >= left){
                left = charIndex[s.charAt(right)] + 1;
            }

            charIndex[s.charAt(right)] = right;
            maxlength = Math.max(maxlength, right - left + 1);
        }

        return maxlength;

    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        maxlength = 0
        left = 0
        charindex = [-1] * 128

        for right in range(n):
            if charindex[ord(s[right])] >= left:
                left = charindex[ord(s[right])] + 1
            charindex[ord(s[right])] = right
            maxlength = max(maxlength, right - left + 1)
        
        return maxlength